### ISSUE TITLE
Considerations for globalSchema.ts

### ISSUE THREAD

**Issue by TonyGravagno** [araera111/mysql-to-zod#156]
The outputToFile function has the name of the globalSchema.ts file hardcoded.

https://github.com/araera111/mysql-to-zod/blob/8cbcf6575542ae3741402376827ea590ae391072/src/process/outputToFile.ts#L37

That can be a problem:

- If, now with the "-f" option, we have two configs that output to the same folder, the one global file will be over-written. This is a problem because the `mysqlTYPE: z.schema()` parameters are written based on what is found in the database. Two different table runs will return different types, and the last run will over-write the previous run, not augment it.
- The new [proposed](https://github.com/TonyGravagno/mysql-to-zod/tree/tg-length-2) maxLength feature creates a function in globalSchema.ts, depending on config settings. If config settings differ, only the last run will determine how that function operates.

While the problem can be easy to solve, choices need to be made about how to solve it. Here are a few ideas, not suggestions, just ideas for how some of this can be addressed:

1. With options.schema.inline=false, create a folder for both the schema.ts and its globalSchema.ts. With the schema.ts importing ".globalSchema", it will be importing its own local global functionality. Example:

```
/schema
. /payables
. . payables.ts
. . globalSchema.ts
. /receivables
. . receivables.ts
. . globalSchema.ts
```

That isn't a great solution. Why have a global file at all when the detail can be in the schema file? The benefit however is that the configuration-specified detail for payables and receivables is not lost. The developer can resolve issues as they wish.

2. Modify the name of the globalSchema file to match the detail : payables.ts, payablesGlobal.ts. Also not a great solution. There is no truly "global" file.

3. Save globalSchema.ts with a JSON object that is written by mysql-to-zod. This is read by every run. It is then merged with the detail for the current run. The updated detail is used to modify the code. So existing parameters of the globalSchema object aren't lost, they are regenerated constantly, augmented with every run.

With that option, differences need to be identified and remedied. For example, if one config says the error message for a field length error is "Too many characters" and another says "Max characters exceeded", this conflict in instructions is an error that needs to be reported when the code is generated. OR code in this area needs to be enhanced to use different messages for different schema (like receivables and payables).

<hr>

Personally I like approach # 3. It's a bit complex for this challenge but I don't think it's too complex. I think it's a decent solution for the problem, without going too far to solve a problem that might not be seen too often.

I would be happy to code that solution as a proposal for this challenge. Maybe that can be just a first attempt to address the issue and if anyone needs a more complex solution we can look at it at that time.

**Comment by araera111** [comment]
Thank you for raising this issue.

As an initial solution, what do you think about users changing the outputDir for each config file? Is the concern that having multiple "global" schemas is problematic? Or is it that you would like to reuse a common schema across different configurations?

Regarding proposal #3, I believe it's feasible. We could start by checking if a globalSchema already exists, load it, and then merge it with the new data while avoiding duplicates. However, I think it might be better to make this an optional feature rather than the default behavior, with the standard being to overwrite the file.

**Comment by TonyGravagno** [comment]
> As an initial solution, what do you think about users changing the outputDir for each config file? Is the concern that having multiple "global" schemas is problematic? Or is it that you would like to reuse a common schema across different configurations?

Yeah, each globalSchema will be slightly different, so it will be up to the developer to make it consistent. So we're just creating a burden for anyone who uses this feature (especially ourselves 😁).

> with the standard being to overwrite the file.

That's the problem now. An application will fail to compile:

- Table 1 only requires mysqlBIGINT.
- Table 2 generated in a second independent run only requires mysqlSTRING.
- globalSchema is overwritten with only mysqlSTRING.
- So the schema.ts for Table 1, referring to the common/global schema, fails to compile because mysqlBIGINT is no longer defined.

With proposal # 3, we read and update globalSchema rather than overwriting.

**Comment by araera111** [comment]
I will take a look at this issue and start working on it. Thank you.

**Comment by araera111** [comment]
I have implemented the new feature as discussed.

- The implementation might be a bit rough at this stage, but for now, I've made sure that keys that are not duplicated are merged.
- I have merged the changes into the develop branch.
- I plan to refine and clean up the source code further.
- When keys are duplicated, I intend to add a message prompting the user to choose which one to use.
- I appreciate your understanding and will continue to improve the implementation.

**Comment by araera111** [comment]
I am confused about the composeGlobalConfig and mergeGlobalConfig functions. I will review the source code to recall how they are implemented.
